### ISSUE TITLE
PostgreSQL taking a long time to start up is considered running and healthy

### ISSUE THREAD

**Issue by ants** [zalando/patroni#283]
When PostgreSQL is taking a long time to start up either due to crash recovery or WAL needed to reach consistent state not being available pg_ctl -w will timeout and return with error code 0.  Postgresql.start() will happily consider such server to be running. This is, at best, misleading.

There seems to be at least one race condition here. 
1. Master boots up and starts Patroni
2. In first iteration of run_cycle Ha.recover() is called, this asynchronously starts postgres.
3. Lock ttl expires.
4. The master is now starting up, will not participate in the race due to async action.
5. Standby will consider itself healthiest because master is not reachable (xlog key missing from JSON) acquires leader key, promotes itself.
6. pg_ctl start timeout expires on old master, it sets role to be replica in Postgresql._do_follow need_rewind == False branch.
7. run_cycle calls follow to restart master with new primary_conninfo while in the middle of recovery, because role != master it will not get rewound.

Basically 3 possible outcomes after a crash
1. Master manages to reacquire the lock in handle_long_action_in_progress in second iteration of run_cycle, will continue to be master once recovery completes. Although it is marked as running when pg_ctl timeout expires, despite not yet accepting connections. Somewhat confusing, but does not require administrator intervention.
2. Lock times out after first but before second iteration, leading to rewind not getting called.
3. Lock times out before first iteration, rewind will get called, everything is peachy.

In some cases (huge checkpoint_timeout, availability is priority over durability) it might be preferrable to refrain from acquiring leader lock before crash recovery is completed.

**Comment by CyberDem0n** [comment]
> In some cases (huge checkpoint_timeout, availability is priority over durability) it might be preferrable to refrain from acquiring leader lock before crash recovery is completed.

Agree. And it looks like the only way to solve it would be having the option which will switch the behaviour.
In one case it should update the lock even before starting dead postgres, in other case it should remove the leader lock immediately, wait couple of seconds (until someboody else will take the lock) and try to rewind and start.
